File: javascript/easy/maxAreaOfIsland.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function(grid) {
    let maxArea = 0, area = 0;
    const dirs = [[0,-1],[0,1],[1,0],[-1,0]];
    for(let i = 0; i < grid.length; i++) {
        for(let j = 0; j < grid[i].length; j++) {
            if(grid[i][j] === 1) {
                area = dfs(grid, i, j, dirs);
                maxArea = Math.max(maxArea, area);
            }
        }
    }
    return maxArea;
};

var dfs = function(grid, i, j, dirs) {
    if(i < 0 || i >= grid.length || j < 0 || j >= grid[i].length || grid[i][j] === 0) {
        return 0;
    }
    grid[i][j] = 0;
    let area = 1;
    for(let dir of dirs) {
        const x = (i + dir[0]), y = (j + dir[1]);
        area += dfs(grid, x, y, dirs);
    }
    
    return area;
}

File: ruby/easy/max_area_of_island.rb
# @param {Integer[][]} grid
# @return {Integer}
def max_area_of_island(grid)
    max_area = 0
    dirs = [[0,1],[0,-1],[1,0],[-1,0]]
    for i in 0..grid.length-1
        for j in 0..grid[i].length-1
            if(grid[i][j] == 1)
                area = dfs(grid, i, j, dirs)
                max_area = [max_area, area].max
            end
        end
    end
    return max_area
end

def dfs(grid, i, j, dirs)
    stack = [[i, j]]
    area = 1
    while !stack.empty?
        c = stack.last
        i, j = c[0], c[1]
        grid[i][j] = 0
        count = 0
        dirs.each do |dir|
            x =(i + dir[0]) and y = (j + dir[1])
            if is_valid(grid, x, y)
                area += 1 and count += 1
                stack.push([x, y])
                break
            end
        end
        stack.pop if count == 0
    end
    return area
end

def is_valid(grid, i, j)
    return i >= 0 && i < grid.length && j >= 0 && j < grid[i].length && grid[i][j] == 1 
end
